base/dates/parse.jl

```diff
@@ -1,131 +1,158 @@
 ### Parsing utilities
 
-function directives{S,F}(::Type{DateFormat{S,F}})
-    tokens = F.parameters
-    di = 1
-    directive_index = zeros(Int, length(tokens))
-    directive_letters = sizehint!(Char[], length(tokens))
-    for (i, token) in enumerate(tokens)
-        if token <: DatePart
-            directive_index[i] = di
-
-            letter = first(token.parameters)
-            push!(directive_letters, letter)
-
-            di += 1
+_directives{S,T}(::Type{DateFormat{S,T}}) = T.parameters
+
+character_codes{S,T}(df::Type{DateFormat{S,T}}) = character_codes(_directives(df))
+function character_codes(directives::SimpleVector)
+    letters = sizehint!(Char[], length(directives))
+    for (i, directive) in enumerate(directives)
+        if directive <: DatePart
+            letter = first(directive.parameters)
+            push!(letters, letter)
         end
     end
-    return tokens, directive_index, directive_letters
+    return letters
 end
 
-genvar(t::DataType) = Symbol(lowercase(string(t.name.name)))
-
-
-@generated function tryparse_core(str::AbstractString, df::DateFormat, raise::Bool=false)
-    token_types, directive_index, directive_letters = directives(df)
+genvar(t::DataType) = Symbol(lowercase(string(Base.datatype_name(t))))
+
+"""
+    tryparsenext_core(str::AbstractString, pos::Int, len::Int, df::DateFormat)
+
+Parses the string according to the directives within the DateFormat. Parsing will start at
+character index `pos` and will stop when all directives are used or we have parsed up to the
+end of the string (`len`).
+
+Returns a 3-element tuple `(values, pos, num_parsed)`:
+* `values::Nullable{Tuple}`: A tuple which contains a values for each `DatePart` within the
+  `DateFormat` in the order in which the occur. If the string ends before we finish parsing
+  all the directives the missing values will be filled in with default values.
+* `pos::Int`: The character index at which parsing stopped.
+* `num_parsed::Int`: The number of values which were parsed and stored within `values`.
+  Useful for distinguishing parsed values from default values.
+"""
+@generated function tryparsenext_core(
+    str::AbstractString, pos::Int, len::Int, df::DateFormat, raise::Bool=false,
```

Comment:
That was an oversight. I made it a keyword at one point and meant to mention it in the docstring details.